presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -2602,9 +2619,26 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
     @Override
-    public List<GrantInfo> listTablePrivileges(ConnectorSession session, SchemaTablePrefix schemaTablePrefix)
+    public Stream<ListTablePrivilegesResult> listTablePrivilegesStream(ConnectorSession session, SchemaTablePrefix prefix)
     {
-        return accessControlMetadata.listTablePrivileges(session, listTables(session, schemaTablePrefix));
+        if (isRedirectToIcebergEnabled(session)) {
```

Comment:
Does this  `if` buy as much?

regardless of the case, we call `listTables(session, prefix)`.
then, in redirect case we call `redirectTable` on each item of the list, which is a very cheap operation if redirect is not enabled.
Thus, you could safely handle both cases with single code path.

Reply:
I don't understand clearly... Can you share a code snippet?

Reply:
```java
@Override
public Stream<ListTablePrivilegesResult> listTablePrivilegesStream(ConnectorSession session, SchemaTablePrefix prefix)
{
    ImmutableList.Builder<SchemaTableName> redirectedTables = ImmutableList.builder();
    ImmutableList.Builder<SchemaTableName> notRedirectedTables = ImmutableList.builder();
    listTables(session, prefix).forEach(table -> {
        if (redirectTable(session, table).isPresent()) {
            redirectedTables.add(table);
        }
        else {
            notRedirectedTables.add(table);
        }
    });
    return Stream.concat(
            accessControlMetadata.listTablePrivileges(session, notRedirectedTables.build()).stream()
                    .map(grantInfo -> new ListTablePrivilegesResult(grantInfo.getSchemaTableName(), Optional.of(grantInfo))),
            redirectedTables.build().stream()
                    .map(table -> new ListTablePrivilegesResult(table, Optional.empty())));
}
```

Reply:
Would that work?